### PR TITLE
Adds RSA-SHA1 method for the OAuth Authentication (issue #8889)

### PR DESCRIPTION
**Pull request by TheFRedFox**
That PR is providing a possible solution for the issue https://github.com/cakephp/cakephp/issues/8889.

It is my first time working with OAuth authentication, so maybe there are still some wrong things.
E.g. I don't know if the consumerSecret and tokenSecret are not required at all for RSA-SHA1 signing.
I know that the token is mandatory for 2-legged Authentication.

Method were added/changed:
- Oauth::authentication()
- Oauth::_rsaSha1()

Tests were added:
- OauthTest::testRsaSigning()
- OauthTest::testRsaSigningWithPassphraseString()
- OauthTest::testRsaSigningWithPassphraseFile()

And some files were added for the tests:
- key.pem (the private key without passphrase)
- key_with_passphrase.pem (the private key with passphrase)
- several key_passphrase with different line endings to support the tests for all operating systems

### CODE REVIEW

**Comment by TheFRedFox** [comment]
Ha, forgot to check phpcs on the OauthTest class. :innocent: Will fix that tomorrow.

Secondly, it seems there is still some failure with the different line endings of the passphrase files. :confused: I will also check that tomorrow.

File: src/Network/Http/Auth/Oauth.php [hunk 1]
```diff
@@ -135,6 +152,64 @@ protected function _hmacSha1($request, $credentials)
     }
 
     /**
+     * Use RSA-SHA1 signing.
+     *
+     * This method is suitable for plain HTTP or HTTPS.
+     *
+     * @param \Cake\Network\Http\Request $request The request object.
+     * @param array $credentials Authentication credentials.
+     * @return string
```
Comment (markstory) [inline src/Network/Http/Auth/Oauth.php]: PHPCS will want an `@throws` doc block tag.

File: src/Network/Http/Auth/Oauth.php [hunk 2]
```diff
@@ -135,6 +152,64 @@ protected function _hmacSha1($request, $credentials)
...
+     */
+    protected function _rsaSha1($request, $credentials)
+    {
+        if (!function_exists('openssl_pkey_get_private')) {
+            throw new \RuntimeException('RSA-SHA1 signature method requires the OpenSSL extension.');
+        }
+
+        $nonce = isset($credentials['nonce']) ? $credentials['nonce'] : uniqid();
```
Comment (markstory) [inline src/Network/Http/Auth/Oauth.php]: I think you want `uniqid(true)`.

File: tests/TestCase/Network/Http/Auth/OauthTest.php [hunk 3]
```diff
@@ -213,4 +213,120 @@ public function testHmacSigning()
...
+     * @return void
+     */
+    public function testRsaSigning() {
+        $request = new Request();
+        $request->url('http://photos.example.net/photos')
+            ->body([
+                       'file' => 'vacaction.jpg',
+                       'size' => 'original'
+                   ]);
+        $private_key_path = TEST_APP . DS . 'config' . DS . 'key.pem';
```
Comment (markstory) [inline tests/TestCase/Network/Http/Auth/OauthTest.php]: Should be `$privateKeyPath`.

File: tests/TestCase/Network/Http/Auth/OauthTest.php [hunk 4]
```diff
@@ -213,4 +213,120 @@ public function testHmacSigning()
...
+     * @return void
+     */
+    public function testRsaSigningWithPassphraseString() {
+        $request = new Request();
+        $request->url('http://photos.example.net/photos')
+            ->body([
+                       'file' => 'vacaction.jpg',
+                       'size' => 'original'
+                   ]);
+        $private_key_path = TEST_APP . DS . 'config' . DS . 'key_with_passphrase.pem';
```
Comment (markstory) [inline tests/TestCase/Network/Http/Auth/OauthTest.php]: Should be `$privateKeyPath`.

File: tests/TestCase/Network/Http/Auth/OauthTest.php [hunk 5]
```diff
@@ -213,4 +213,120 @@ public function testHmacSigning()
...
+
+        if(PHP_EOL == "\n") $passphrase_path = TEST_APP . DS . 'config' . DS . 'key_passphrase_lf';
+        else if(PHP_EOL == "\r\n") $passphrase_path = TEST_APP . DS . 'config' . DS . 'key_passphrase_crlf';
+        else if(PHP_EOL == "\r") $passphrase_path = TEST_APP . DS . 'config' . DS . 'key_passphrase_cr';
+        else { $this->markTestSkipped('The file for the key passphrase could not be loaded as PHP_EOL could not be recognized.'); return; }
```
Comment (markstory) [inline tests/TestCase/Network/Http/Auth/OauthTest.php]: Can we just use `\n`. Doing platform dependent line endings is going to make deploying windows built code/files to a nix server 'work in dev, but not prod'. If we always use consistent line endings the pem file will either work, or not consistently.

File: src/Network/Http/Auth/Oauth.php (hunk 2 above)
Comment (TheFRedFox) [inline src/Network/Http/Auth/Oauth.php]: I copied the nonce default creation from the `_hmacSha1` method. Should it be `uniqid(true)` in this method too? The `_plaintext` method also just has `uniqid()`.
Btw. we can also use `Security::randomBytes()` like in #8890. :wink:

File: src/Network/Http/Auth/Oauth.php (hunk 1 above)
Comment (TheFRedFox) [inline src/Network/Http/Auth/Oauth.php]: Actually PHPCS didn't complain about it. But sure, I will add it.

File: tests/TestCase/Network/Http/Auth/OauthTest.php (hunk 5 above)
Comment (TheFRedFox) [inline tests/TestCase/Network/Http/Auth/OauthTest.php]: Therefore my idea was to provide three different files with three different line endings, so the tests can always run. I guess it failed, because while pushing the files my git client automatically convert the file. But it would also be ok with me to make the test skip if PHP_EOL is anything but `\n`.

File: src/Network/Http/Auth/Oauth.php (hunk 2 above)
Comment (markstory) [inline src/Network/Http/Auth/Oauth.php]: :+1: `Security::randomBytes()` is probably better.

File: tests/TestCase/Network/Http/Auth/OauthTest.php (hunk 5 above)
Comment (TheFRedFox) [inline tests/TestCase/Network/Http/Auth/OauthTest.php]: Ah, I just remember that git clients also may convert the line endings while pulling/cloning. I will change it then, that just `\n` is supported, otherwise the test will be skipped.

**Comment by TheFRedFox** [comment]
I changed the `uniqid()` to `Security::randomBytes(16)`. Is length of 16 enough, or should it be more?

I removed the other passphrase files again, and I really hope the line endings were pushed correctly this time. ^^

**Comment by markstory** [comment]
> Is length of 16 enough

That should be plenty. Its a nonce :)

File: src/Network/Http/Auth/Oauth.php [hunk 6]
```diff
@@ -54,10 +49,33 @@ public function authentication(Request $request, array $credentials)
...
+                );
+                if (!$hasKeys) {
+                    return;
+                }
                 $value = $this->_hmacSha1($request, $credentials);
                 break;
 
+            case 'RSA-SHA1':
+                if (!isset($credentials['privateKeyFile'])) {
+                    return;
```
Comment (markstory) [inline src/Network/Http/Auth/Oauth.php]: We should throw an exception here. The developer has made a bad mistake.

File: tests/TestCase/Network/Http/Auth/OauthTest.php [hunk 7]
```diff
@@ -213,4 +213,125 @@ public function testHmacSigning()
...
+        $request->url('http://photos.example.net/photos')
+            ->body([
+                       'file' => 'vacaction.jpg',
+                       'size' => 'original'
+                   ]);
+        $privateKeyPath = TEST_APP . DS . 'config' . DS . 'key_with_passphrase.pem';
+
+        if (PHP_EOL != "\n") {
+            $this->markTestSkipped('Just the line ending ("\n") is supported. You can run the test again e.g. on a linux system.');
+            return;
```
Comment (markstory) [inline tests/TestCase/Network/Http/Auth/OauthTest.php]: You don't need a return here. `markTestSkipped()` throws an exception.

File: src/Network/Http/Auth/Oauth.php (hunk 6 above)
Comment (TheFRedFox) [inline src/Network/Http/Auth/Oauth.php]: But I think this would be inconsistent to the other checks. It is also a bad mistake if the developer forgot the consumerKey. I would then suggest to consistently throw an exception whenever a mandatory credential key is missed.

File: tests/TestCase/Network/Http/Auth/OauthTest.php (hunk 7 above)
Comment (TheFRedFox) [inline tests/TestCase/Network/Http/Auth/OauthTest.php]: Well, I actually knew that .. I like the `return` there because the developer or rather I can clearly see that it will definitely end there (and also it is also easier for the IDE).
But nevertheless, if you guys never use `return` there I will remove it again. :wink:

File: src/Network/Http/Auth/Oauth.php (hunk 6 above)
Comment (markstory) [inline src/Network/Http/Auth/Oauth.php]: Good point. We can approach that problem separately.

File: tests/TestCase/Network/Http/Auth/OauthTest.php (hunk 7 above)
Comment (markstory) [inline tests/TestCase/Network/Http/Auth/OauthTest.php]: We also have a `skipIf()` helper, that takes a boolean, and conditionally skips the test.

```php
$this->skipIf(PHP_EOL != "\n", 'Just the line ending ("\n") is supported. You can run the test again e.g. on a linux system.');
```
could be used.

File: src/Network/Http/Auth/Oauth.php (hunk 6 above)
Comment (TheFRedFox) [inline src/Network/Http/Auth/Oauth.php]: Sry, but you confused me a bit now. :smile: You mean, we can do it in another PR? So we can finish this PR now? ^^

File: src/Network/Http/Auth/Oauth.php [hunk 8]
```diff
@@ -135,6 +153,66 @@ protected function _hmacSha1($request, $credentials)
...
+        $credentials += [
+            'privateKeyPassphrase' => null,
+        ];
+        if (is_resource($credentials['privateKeyPassphrase'])) {
+            $resource = $credentials['privateKeyPassphrase'];
+            $passphrase = stream_get_line($resource, 0, PHP_EOL);
+            rewind($resource);
+            $credentials['privateKeyPassphrase'] = $passphrase;
+        }
+        $privateKey = openssl_pkey_get_private(file_get_contents($credentials['privateKeyFile']), $credentials['privateKeyPassphrase']);
```
Comment (TheFRedFox) [inline src/Network/Http/Auth/Oauth.php]: Do we want to change `privateKeyFile` to `privateKey` and to support `string` and `file` representations of the private key? One more opportunity for the developer to pass the private key.
And it would be consistent to the privateKeyPassphrase option (`string` and `file` representation).

File: src/Network/Http/Auth/Oauth.php (hunk 6 above)
Comment (markstory) [inline src/Network/Http/Auth/Oauth.php]: Yes, a separate pull request at some point in the future.

File: src/Network/Http/Auth/Oauth.php (hunk 8 above)
Comment (markstory) [inline src/Network/Http/Auth/Oauth.php]: :+1: to consistency. Having one option that reads resources/strings means we can avoid the awkward scenario where both are defined.

**Comment by TheFRedFox** [comment]
Alright. Made it consistent to the passphrase parameter now and added three more test to check this functionality.
I guess I am finished now. ^^
